src/RA_httpthread.cpp

```diff
@@ -252,31 +247,74 @@ void RAWeb::SetUserAgentString()
+                        0ULL, VER_MAJORVERSION, VER_GREATER_EQUAL),
+                        VER_MINORVERSION, VER_GREATER_EQUAL),
+                        VER_SERVICEPACKMAJOR, VER_GREATER_EQUAL)
+                };
+
+                return VerifyVersionInfoW(&osvi, VER_MAJORVERSION | VER_MINORVERSION |
+                                          VER_SERVICEPACKMAJOR, dwlConditionMask) != FALSE;
+            };
+            if (!IsXPOrGreater())
+                return;
+        }
+
+        using fnRtlGetVersion = NTSTATUS(NTAPI*)(PRTL_OSVERSIONINFOEXW lpVersionInformation);
+
+        const auto& ntModule{ ::GetModuleHandleW(L"ntdll.dll") };
+        if (!ntModule)
+            return;
+
+        RTL_OSVERSIONINFOEXW osVersion{};
+        const auto& RtlGetVersion{
+            reinterpret_cast<fnRtlGetVersion>(::GetProcAddress(ntModule, "RtlGetVersion"))
+        };
+        if (!RtlGetVersion)
+            return;
         RtlGetVersion(&osVersion);
-        if (osVersion.dwMajorVersion > 0)
+        
         {
-            sUserAgent.append("WindowsNT ");
-            sUserAgent.append(std::to_string(osVersion.dwMajorVersion));
-            sUserAgent.append(".");
-            sUserAgent.append(std::to_string(osVersion.dwMinorVersion));
+            std::ostringstream oss;
+            oss << "WindowsNT " << osVersion.dwMajorVersion << '.' << osVersion.dwMinorVersion
+                << '.' << osVersion.dwBuildNumber;
+            sUserAgent.append(oss.str());
         }
     }
 
     sUserAgent.append(") Integration/");
+    {
+        std::ostringstream oss;
+        // Compile-time string concatenation will need some user-defined stuff/work.
+        oss << "WindowsNT " << RA_INTEGRATION_VERSION_MAJOR << '.' << RA_INTEGRATION_VERSION_MINOR
+            << '.' << RA_INTEGRATION_VERSION_REVISION << '.' << RA_INTEGRATION_VERSION_MODIFIED;
+        sUserAgent.append(oss.str());
+    }
 
-    auto buffer{ std::make_unique<char[]>(64U) };
-    sprintf_s(buffer.get(), 64U, "%d.%d.%d.%d", RA_INTEGRATION_VERSION_MAJOR, RA_INTEGRATION_VERSION_MINOR,
-              RA_INTEGRATION_VERSION_REVISION, RA_INTEGRATION_VERSION_MODIFIED);
-    sUserAgent.append(buffer.release());
-
-    const char* ptr = strchr(RA_INTEGRATION_VERSION_PRODUCT, '-');
-    if (ptr != nullptr)
-        sUserAgent.append(ptr);
+    {
+        // compile-time finding, intellisense will tell you the number is already
+        // Only put C-Strings in string_view, they must also live longer than the string_view
+        _CONSTANT_LOC posFound{ std::string_view{ RA_INTEGRATION_VERSION_PRODUCT }.find('-') };
+        if constexpr (posFound != std::string_view::npos)
+        {
+            // appending still done at runtime but the values for it are retrieved at compile-time
+            // needs to use the const T& overload
+            std::string_view sAppend{ RA_INTEGRATION_VERSION_PRODUCT };
+            sUserAgent.append(sAppend, posFound);
+        }
```

Comment:
I like the idea of resolving this at compile time, but how is it related to the PCH?

Reply:
It's not, I think too far ahead sometimes. I'll try not to do it again.